stored_requests/ab_fetcher.go

```diff
@@ -0,0 +1,212 @@
+// and returns the randomly selected ABConfig (undecoded)
+func runABSelection(jsonBytes []byte, keys ...string) (abConfig *ABConfig, err error) {
+	var abConfigJSON []byte
+	throw := 100 * rand.Float64()
+	t := 0.0
+	_, e := jsonparser.ArrayEach(jsonBytes, func(value []byte, dataType jsonparser.ValueType, offset int, e error) {
+		if err != nil || e != nil || len(abConfigJSON) > 0 {
+			return
+		}
+		if percent, err := jsonparser.GetFloat(value, "ratio"); err == nil {
+			t += percent
+			if throw < t {
+				abConfigJSON = value
+			}
+		}
+	}, keys...)
+	if e != nil {
+		return nil, e
+	}
+	if err != nil {
+		return nil, err
+	}
+	if t > 100 {
+		return nil, fmt.Errorf(`%v sum of probabilities %.1f greater than 100%%`, keys, t)
+	}
+	if len(abConfigJSON) == 0 {
+		return nil, nil // control set
+	}
+	return unmarshalABConfig(abConfigJSON)
+}
+
+// unmarshalABConfig parses and validates ABConfig. Could also use the slower json.UnMarshal instead.
+func unmarshalABConfig(abConfigJSON []byte) (*ABConfig, error) {
+	// Parse and validate ABConfig. Could also use the slower json.UnMarshal instead.
+	abConfig := ABConfig{}
+	var err error
+	abConfig.Code, err = jsonparser.GetString(abConfigJSON, "code")
+	if err != nil {
+		return nil, fmt.Errorf(`parsing "code" in ab_config: %v`, err)
+	}
+	if len(abConfig.Code) == 0 {
+		return nil, fmt.Errorf(`"code" tag is required in all ab_config sections`)
+	}
+	abConfig.RequestID, _ = jsonparser.GetString(abConfigJSON, "request_id") // request_id can be missing, ignore error
+	abConfig.ImpIDs = make(map[string]string)                                // imp_ids can be missing, so ignore error
+	_ = jsonparser.ObjectEach(abConfigJSON,
+		func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+			if len(value) > 0 && len(key) > 0 {
+				abConfig.ImpIDs[string(key)] = string(value)
+			}
+			return nil
+		},
+		"imp_ids")
+	if len(abConfig.RequestID) == 0 && len(abConfig.ImpIDs) == 0 {
+		return nil, fmt.Errorf(`no replacement ids specified in ab_config for code="%s" `, abConfig.Code)
+	}
+	return &abConfig, nil
+}
+
+/* ABConfig defines an element such as in the list below:
+[
+	{
+		"code": "1321312313312312",
+		"ratio": 15.0,
+		"request_id": "ABCD-0123-4567-111111", ⬅ this is the replacement stored request id
+		"imp_ids": {
+			"DCBA-3333-4444-012345": "DCBA-3333-4444-111111", ⬅ replacement stored imp id
+			"DCBA-4444-5555-666666": "DCBA-4444-9999-222222", ⬅ replacement stored imp id
+			...
+		}
+	},
+	...
+]
+*/
+type ABConfig struct {
+	// Code is a string identifying this experiment.
+	// The control set must not be specified (its ratio inferred and it makes no replacements)
+	Code string `json:"code"`
+	// Ratio is the percentage of requests that should have this experiment applied (0-100)
+	// The sum of Ratio values for all the experiments in the group must be <100
+	Ratio float64 `json:"ratio"`
```

Comment:
Needing the ratios to sum to 100% might become annoying and error prone. What do you think of a "Weight" instead of a "Ratio". Where the ratio is computed as ABConfig.Weight / Sum Of All ABConfig.Weights?

It would still work the same if someone wants to sum up to 100, but otherwise would gracefully handle other situations.

Reply:
We want to control the exact ratio for analytics. With weights, we have the opposite problem that adding a weight changes the distribution of the rest.

The 100% requirement shouldn't be a problem, because most of the time tests involve a fraction of the traffic, leaving at least 80-90% unchanged.